rfd/0154-logging-guidelines.md

```diff
@@ -0,0 +1,278 @@
+to the configured output can be used.
+
+Once a `slog.Logger` may be retrieved either globally via `slog.Default`/`slog.With` or by dependency injection, the
+process of converting individual loggers may begin. Since `TeleportProcess` is the entry point and object responsible
+for creating other components it should be the first component to be converted to have and use a `slog.Logger`. Once it
+has been converted, objects it creates can be migrated independently until they've all been converted. The `api`module
+consumes the global `logrus` logger, so it may be converted independently after the global `slog.Logger` is configured
+in the first step. However, the minimum supported go version of the `api` module will need to be bumped to >= go 1.21
+before package `log/slog` is available to be consumed there. When the last consumers of the `logrus` library have been
+removed from the `api` module `go mod tidy` to get rid of the `logrus` dependency (or at least make it an indirect
+dependency). Once `slog` is the only library writing output from within `teleport` the log initialization logic can be
+cleaned up to only create a `slog.Logger` and remove the `io.Writer` wrapper from the first step. At this
+point `go mod tidy` should be run to clean up the dependency. A depguard rule should also be put in place to prevent any
+accidental uses of `logrus` after the conversion.
+
+### API Differences
+
+There are a few subtle API differences between the two logging libraries which prevent migrating to be a find and
+replace operation. The following sections demonstrate these differences and offer various recommendations to overcome
+them when migrating to `log/slog`.
+
+#### Formatted messages
+
+`logrus` provides an API which allows for `fmt.Sprintf` style formatting (Tracef,Debugf,Infof, etc.) of log messages
+which are relied on heavily throughout the codebase.
+
+> ```go
+> logrus.Infof("Connected to cluster %v at %v", clusterName, conn.RemoteAddr())
+>```
+>```bash
+>2023-11-02T15:18:40-04:00 [PROXY]    INFO Connected to cluster foo at 192.168.1.243. pid:14968.1  regular/proxy.go:250
+>```
+
+The `log/slog` API differs in that all logging functions provided take a message and a series of key value pairs,
+referred to as attributes, instead of a message to format and associated arguments. Any attributes provided will be
+included as fields alongside the message. While a literal translation of the above call to `Infof` is possible in slog
+by using `fmt.Sprintf` directly it is not recommended. Instead, when converting to `log/slog` structured logging should
+be embraced. The message from above can be translated to carry a bit more context, and the additional fields in the
+message can be moved to the attributes.
+
+> ```go
+> slog.Info("Proxied connection to cluster established.", "cluster_name", clusterName, "remote_address", conn.RemoteAddr())
```

Comment:
Is this a good place to discuss full stop or not for one sentence logging statements?

Reply:
Yes, but full stop at end or not?

Reply:
Yes, sorry if it wasn't clear.

Reply:
An explicit suggestion would be great, thanks.